core/src/main/java/tech/ydb/core/impl/pool/EndpointPool.java

```diff
@@ -66,48 +89,65 @@ public EndpointRecord getEndpoint(@Nullable Integer preferredNodeID) {
         }
     }
 
-    private int getStartPriority(String selfLocation, String location) {
-        if (balancingSettings.getPolicy() == BalancingSettings.Policy.USE_ALL_NODES) {
-            return 0;
-        }
-
-        String prefered = balancingSettings.getPreferableLocation();
-        if (prefered == null || prefered.isEmpty()) { // If prefered location were not specified - use self location
-            prefered = selfLocation;
-        }
-
-        return (location != null && location.equalsIgnoreCase(prefered)) ? 0 : LOCALITY_SHIFT;
-    }
-
     // Sets new endpoints, returns removed
     public List<EndpointRecord> setNewState(DiscoveryProtos.ListEndpointsResult result) {
         String selfLocation = result.getSelfLocation();
 
         Set<String> newKnownEndpoints = new HashSet<>();
         Map<Integer, PriorityEndpoint> newKnownEndpointsByNodeId = new HashMap<>();
         List<PriorityEndpoint> newRecords = new ArrayList<>();
+        Map<String, List<DiscoveryProtos.EndpointInfo>> dcLocationToNodes = result
+                .getEndpointsList()
+                .stream()
+                .collect(Collectors.groupingBy(DiscoveryProtos.EndpointInfo::getLocation));
 
         logger.debug("init new state with location {} and {} endpoints", selfLocation, result.getEndpointsCount());
-        for (DiscoveryProtos.EndpointInfo info : result.getEndpointsList()) {
-            int priority = getStartPriority(selfLocation, info.getLocation());
-            PriorityEndpoint entry = new PriorityEndpoint(info, priority);
-            String endpoint = entry.getHostAndPort();
-
-            if (!newKnownEndpoints.contains(endpoint)) {
-                logger.debug("  added endpoint {}", entry);
-                newKnownEndpoints.add(endpoint);
-                newKnownEndpointsByNodeId.put(entry.getNodeId(), entry);
-                newRecords.add(entry);
-            } else {
-                logger.warn("dublicate endpoint {}", entry.getHostAndPort());
-            }
-        }
+
+        dcLocationToNodes.forEach(
+                (location, nodes) -> {
+                    int nodeSize = Math.min(NODE_SIZE, nodes.size());
+
+                    long priorityNodesFromLocation = nodes
```

Comment:
I think we cannot use it, because we have to detect the local data center, not nearest endpoint. We have to use all endpoints from the local DC, but here we just sort endpoints by tcp ping durations.